contributors/design-proposals/watchable-discovery.md

```diff
@@ -0,0 +1,231 @@
+	// Clients can use the appropriate server address as per the CIDR that they match.
+	// In case of multiple matches, clients should use the longest matching CIDR.
+	// The server returns only those CIDRs that it thinks that the client can match.
+	// For example: the master will return an internal IP CIDR only, if the client reaches the server using an internal IP.
+	// Server looks at X-Forwarded-For header or X-Real-Ip header or request.RemoteAddr (in that order) to get the client IP.
+	ServerAddressByClientCIDRs []metav1.ServerAddressByClientCIDR `json:"serverAddressByClientCIDRs" protobuf:"bytes,2,rep,name=serverAddressByClientCIDRs"`
+}
+
+// GroupVersion describes one version of an API group, including all resources available in this version.
+type GroupVersion struct {
+	// the version name
+	Name string `json:"name" protobuf:"bytes,1,opt,name=name"`
+	// resources contains the name of the resources and if they are namespaced.
+	Resources []metav1.APIResource `json:"resources" protobuf:"bytes,2,rep,name=resources"`
+}
+
+// GroupList is a list of API groups.
+type GroupList struct {
+	metav1.TypeMeta `json:",inline"`
+	// +optional
+	metav1.ListMeta `json:"metadata,omitempty" protobuf:"bytes,1,opt,name=metadata"`
+
+	// List of group.
+	Items []Group `json:"items" protobuf:"bytes,2,rep,name=items"`
+}
+```
+
+## Behavior
+
+Each API server in the cluster serves its own `discovery.k8s.io` API
+group. By default, this is backed by normal etcd storage, i.e. the
+semantics of the `ResourceVersion` is the one of the underlaying etcd
+storage backend.
+
+Each API server has its own `DiscoveryUpdateController` that's
+regularly reconciles the storage state of the discovery API group with
+its internal registered/announce API group registry. Moreover, when
+a new API groups is installed or removed in its HTTP handler, the
+corresponding API group object is updated synchronously.
+
+### Aggegration
+
+The aggregator watches the discovery API group of each of its proxied
+API servers. It filters the returned discovery information by taking
+the `APIService` status into consideration, i.e. by 
+- only showing those GroupVersions that are registered via an
+  `APIService`
+- only showing those GroupVersions that are accepted by the aggregator
+  (i.e. which have no naming conflict).
+
+The aggregator will update its own discovery API group resources
+triggered by changes in the sub-API-servers.
+
+### Embedded Aggregator in kube-apiserver
+
+In this mode, both the aggregator and the kube-apiserver share the
+storage. To avoid conflicts about updating the API group, the `Group`
+objects of kube-apiserver are marked as "local" with an
+annotation. The aggregator will not touch them and the other way
+around for the kube-apiserver for the not-marked `Group` objects.
+
+### HA
+
+API servers in HA mode will share a storage. Hence, they will share
+the state of the discovery API group. Their controllers will fight for
```

Comment:
Yeah this demonstrates why the plan of each of them doing its own thing is not good. (honestly when reading the previous section I thought you were going to have each apiserver store the data in a separate path to avoid this collision)